File: network/kmm_network/androidNetworkApp/src/main/java/com/example/kmm_network/android/presentation/user_list/UserListViewModel.kt
package com.example.kmm_network.android.presentation.user_list

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.kmm_network.presentation.user_list.UserListState
import com.example.kmm_network.domain.model.User
import com.example.kmm_network.interactors.CreateUser
import com.example.kmm_network.interactors.GetUserList
import com.example.kmm_network.presentation.user_list.UserListEvents
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import javax.inject.Inject

@HiltViewModel
class UserListViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle, // we don't need it here, but hilt requires it?
    private val getUserList: GetUserList,
) : ViewModel() {

    val state: MutableState<UserListState> = mutableStateOf(UserListState())

    init {
        onTriggerEvent(UserListEvents.LoadUsers)
    }

    fun onTriggerEvent(event: UserListEvents) {
        when (event) {
            UserListEvents.LoadUsers -> {
                loadUsers()
            }
            UserListEvents.NextPage -> {
                nextPage()
            }
            UserListEvents.Refresh -> {
                refresh()
            }
            else -> {
                println("### some event")
            }
        }
    }

    private fun refresh() {
        state.value = state.value.copy(page = 1)

        getUserList.executeJustCache(
            page = state.value.page
        ).onEach { dataState ->

            state.value = state.value.copy(isLoading = dataState.isLoading)

            dataState.data?.let { users ->
                state.value = state.value.copy(users = users)
            }

            dataState.message?.let { message ->
                println(message)
            }
        }.launchIn(viewModelScope)
    }

    private fun loadUsers() {
        getUserList.execute(
            page = state.value.page
        ).onEach { dataState ->

            state.value = state.value.copy(isLoading = dataState.isLoading)

            dataState.data?.let { users ->
                appendUser(users = users)
            }

            dataState.message?.let { message ->
                println(message)
            }
        }.launchIn(viewModelScope)
    }

    private fun nextPage() {
        state.value = state.value.copy(page = state.value.page + 1)
        loadUsers()
    }

    private fun appendUser(users: List<User>) {
        val curr = ArrayList(state.value.users)
        curr.addAll(users)
        state.value = state.value.copy(users = curr)
    }

}

File: network/kmm_network/iosNetworkApp/iosNetworkApp/presentation/user_list/UserListViewModel.swift
import SwiftUI
import shared

class UserListViewModel: ObservableObject {
    
    let getUserList: GetUserList
    
    @Published var state: UserListState = UserListState(isLoading: false, page: 1, users: [User]())

    @Published var lastEmailInTheList: String = ""
    
    init(getUserList: GetUserList) {
        self.getUserList = getUserList
        loadUserList()
    }
    
    func updateState(
        isLoading: Bool? = nil,
        page: Int? = nil
    ) {
        let currentState = (self.state.copy() as! UserListState)
        self.state = self.state.doCopy(
            isLoading: isLoading ?? currentState.isLoading,
            page: Int32(page ?? Int(currentState.page)),
            users: currentState.users
        )
    }
    
    func loadUserList() {
        let currentState = (self.state.copy() as! UserListState)
        getUserList.execute(
            page: Int32(currentState.page)
        ).collectCommon(
            coroutineScope: nil,
            callback: { dataState in
                if dataState != nil {
                    let data = dataState?.data
                    let message = dataState?.message
                    let loading = dataState?.isLoading ?? false
                    
                    self.updateState(isLoading: loading)
                    
                    if data != nil {
                        self.appendUserList(userList: data as! [User])
                    }
                    
                    if message != nil {
                        print("Error message: \((message ?? "NON ERROR") as String)")
                    }
                }
            }
        )
    }
    
    func refresh() {
        getUserList.executeJustCache(
                page: 1
            ).collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        let loading = dataState?.isLoading ?? false
                        
                        self.updateState(isLoading: loading)
                        
                        if data != nil {
                            self.state = self.state.doCopy(
                                isLoading: loading,
                                page: 1,
                                users: data as! [User]
                            )
                            self.lastEmailInTheList = (data as! [User]).last?.email ?? ""
                        }
                        
                        if message != nil {
                            print("Error message: \((message ?? "NON ERROR") as String)")
                        }
                    }
                }
            )
    }
    
    func appendUserList(userList: [User]) {
        let currentState = (self.state.copy() as! UserListState)
        let newUserList = currentState.users + userList
        self.state = self.state.doCopy(
            isLoading: currentState.isLoading,
            page: currentState.page,
            users: newUserList
        )
        self.lastEmailInTheList = newUserList.last?.email ?? ""
    }
    
    func shouldQueryNextUserPage(user: User) -> Bool {
        if ((user.email == self.lastEmailInTheList)
            && isCurrentPageTooBig()
            && !self.state.isLoading)
        {
            return true
        }
        return false
    }
    
    private func isCurrentPageTooBig() -> Bool {
        return 6 * self.state.page <= self.state.users.count
    }
    
    func nextPage() {
        let currentSatate = (self.state.copy() as! UserListState)
        self.updateState(page: Int(currentSatate.page) + 1)
        self.loadUserList()
    }
}
